File: android/app/src/main/java/com/foliole/android/FolioleCompanionCanonicalAttachmentKey.java
package com.foliole.android;

import java.util.HashMap;
import java.util.Map;

final class FolioleCompanionCanonicalAttachmentKey {
    private static final Map<String, String> EXTENSIONS = new HashMap<>();
    static {
        EXTENSIONS.put("application/pdf", ".pdf");
        EXTENSIONS.put("application/epub+zip", ".epub");
        EXTENSIONS.put("image/gif", ".gif");
        EXTENSIONS.put("image/jpeg", ".jpg");
        EXTENSIONS.put("image/png", ".png");
        EXTENSIONS.put("image/webp", ".webp");
    }

    private FolioleCompanionCanonicalAttachmentKey() {}

    static String storageKey(String contentHash, String mimeType) {
        String extension = EXTENSIONS.get(mimeType == null ? null : mimeType.trim().toLowerCase());
        return extension != null && contentHash != null && contentHash.matches("[a-f0-9]{64}")
            ? contentHash + extension : null;
    }

    static boolean valid(String key) {
        return key != null && key.length() > 64 && key.substring(0, 64).matches("[a-f0-9]{64}")
            && EXTENSIONS.containsValue(key.substring(64));
    }

    static boolean matches(String contentHash, String mimeType, String storageKey) {
        return storageKey != null && storageKey.equals(storageKey(contentHash, mimeType));
    }
}

File: ios/App/App/FolioleCompanionCanonicalAttachmentKey.swift
import Foundation

enum FolioleCompanionCanonicalAttachmentKey {
    private static let extensions = [
        "application/epub+zip": ".epub", "application/pdf": ".pdf", "image/gif": ".gif", "image/jpeg": ".jpg",
        "image/png": ".png", "image/webp": ".webp"
    ]

    static func storageKey(contentHash: String, mimeType: String) -> String? {
        guard contentHash.range(of: "^[a-f0-9]{64}$", options: .regularExpression) != nil,
              let suffix = extensions[mimeType.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()] else {
            return nil
        }
        return contentHash + suffix
    }

    static func valid(_ key: String) -> Bool {
        guard key.count > 64, String(key.prefix(64)).range(of: "^[a-f0-9]{64}$", options: .regularExpression) != nil else { return false }
        return extensions.values.contains(String(key.dropFirst(64)))
    }

    static func matches(contentHash: String, mimeType: String, storageKey: String) -> Bool {
        return storageKey == self.storageKey(contentHash: contentHash, mimeType: mimeType)
    }
}
